Make OusterDriver a class, not a class template

rotu opened the pull request:
Fixes the build under Clang
Supersedes #27
Fixes #13

rotu committed:
Make OusterDriver a class, not a class template

Fixes the build under Clang
Supersedes #27

SteveMacenski commented on ros2_ouster/include/ros2_ouster/ouster_driver.hpp:
```diff
@@ -51,7 +53,9 @@ class OusterDriver : public lifecycle_interface::LifecycleInterface
    * @param options Node options for lifecycle node interfaces
    */
-  explicit OusterDriver(const rclcpp::NodeOptions & options);
+  OusterDriver(
```
That is non permissible with the composition API, you cannot change the constructor signature.

rotu replied:
Is there a test case or documentation? Also, why is it commented out? https://github.com/SteveMacenski/ros2_ouster_drivers/blob/f3bc817f21212c357815e3d0b468ddbe345ffb40/ros2_ouster/include/ros2_ouster/driver_types.hpp#L33

SteveMacenski replied:
Lifecycle and components are mutually exclusive at this moment. Also had some issues with templates and components but stopped before debugging knowing the lifecycle issue.

rotu replied:
So what do you mean by “composition API”? Do you mean RCLCPP_COMPONENTS_REGISTER_NODE or something else?

SteveMacenski replied:
Using components manager

SteveMacenski commented on ros2_ouster/src/OS1/OS1_sensor.cpp:
```diff
@@ -22,18 +22,11 @@ namespace OS1
-}
+: SensorInterface(),
+  _lidar_packet(lidar_packet_bytes + 1),
+  _imu_packet(imu_packet_bytes + 1) {}
```
Please add brackets on new lines

rotu committed:
Code review

1. Created a concrete OS1Driver class compatible with RCLCPP_COMPONENTS_REGISTER_NODE
2. Uncrustify
3. Remove driver_types.hpp and add some missing headers to ouster_driver.cpp. Note the accidental dependence on processor_factories.hpp which should probably be refactored.

rotu commented:
Back to you, @SteveMacenski.

SteveMacenski commented on ros2_ouster/src/main.cpp:
```diff
@@ -13,15 +13,14 @@
   rclcpp::init(argc, argv);
   auto options = rclcpp::NodeOptions();
-  auto node = std::make_shared<ros2_ouster::OS1Driver>(options);
+  auto node = std::make_unique<ros2_ouster::OS1Driver>(options);
```
This must remain shared in order to use `shared_from_this` otherwise this will seg fault.

rotu replied:
Oof you’re totally right!!! I need to fix this

SteveMacenski commented on ros2_ouster/src/OS1/OS1_sensor.cpp:
```diff
@@ -22,18 +22,12 @@ namespace OS1
 
-OS1Sensor::~OS1Sensor()
-{
-  _ouster_client.reset();
```
Please re-add these clears.

rotu replied:
Why? These are all the same as would be provided in the default destructor, and this way you don’t have to read the destructor to know that.

SteveMacenski replied:
Deterministic shutdown. If the packets are reset before the client, it can segfault and crash on shutdown. Also so that you can inherit from this class and control which resources get deallocated when.

rotu replied:
Fixed. If you really want the explicit destructor, you can have it, but the possible crash here tells me that those vectors should be internal to the `client` type.

rotu replied:
reverted this change. It's inessential to the PR and many other destructors should be cleaned up as well.

SteveMacenski commented on ros2_ouster/include/ros2_ouster/driver_types.hpp:
```diff
@@ -14,22 +14,18 @@
 
-template class ros2_ouster::OusterDriver<OS1::OS1Sensor>;
-using OS1Driver = ros2_ouster::OusterDriver<OS1::OS1Sensor>;
+class OS1Driver : public OusterDriver
```
This seems hacky and not really in line with how anyone would expect this to be formatted. 

What's the actual goal of this PR? It seems like your just shuffling around code at this point. I thought we discussed in the other PR that we just needed to clean up who included which header to result the warning.

rotu replied:
There is a method to my madness! This is needed for the component macro so (1) the class has a constructor with a single NodeOptions argument (2) the `RCLCPP_COMPONENTS_REGISTER_NODE` macro appears in a single translation unit as per its documentation. In a header, it would be in every translation unit that includes that header.

SteveMacenski replied:
How would that generalize if we have multiple classes of driver types in that same file? 

I don't like the idea of having these shim classes. If we're gun-hoe on using the interfaces maybe we should just instantiate them in the driver itself. ~Another option is to have a `setDevice()` function~ which doesn't work well if its being auto-configured on launch, I don't believe (not 100% sure)

rotu replied:
I don’t love it either. But it can be turned into a template *really* easily and seems simpler than adjusting the component infrastructure to work with a factory method. Just think of it as an explicit template instantiation, which it basically is.

SteveMacenski replied:
Let me think about this more. I dont like the shim + interface, it feels very template-y but without the power of templates

rotu replied:
and without the lack of control over where/whether your template class is instantiated. You need to instantiate the class for the component interface, so IMO nothing lost here.

SteveMacenski commented on ros2_ouster/src/driver_types.cpp:
```diff
@@ -0,0 +1,26 @@
+#include "ros2_ouster/OS1/OS1_sensor.hpp"
+
+ros2_ouster::OS1Driver::OS1Driver(rclcpp::NodeOptions options)
+: OusterDriver{std::make_unique<OS1::OS1Sensor>(), options}
```
You instantiate a unique pointer then immediately move it, you really should just make the pointer in the driver at that point or just hand it a reference to an object or a raw pointer.

rotu replied:
The unique pointer move will be elided. A reference or a pointer would not have the same ownership semantics. Would you be happier with an rvalue reference?

SteveMacenski replied:
Some of this mechanics is why templates where used to begin with so there wouldn't be a need for a bunch of shims. Perhaps the right solution is to remove the interfaces and use only the template.

SteveMacenski replied:
Also place this in a namespace, don't append the constructor with `ros2_ouster`. It makes it hard to modify.

SteveMacenski reviewed (changes requested):
I'm not sure the path forward on this PR. It seems like its just shuffling around code for the sake of it rather than containing some superior structure

SteveMacenski commented on ros2_ouster/src/driver_types.cpp:
```diff
@@ -0,0 +1,26 @@
+: OusterDriver{std::make_unique<OS1::OS1Sensor>(), options}
+{}
+
+// RCLCPP_COMPONENTS_REGISTER_NODE(ros2_ouster::OS1Driver)
```
Did you test that this actually fixes the components issue?

rotu replied:
No. I just made it follow the documented usage of RCLCPP_COMPONENTS_REGISTER_NODE. What was your components issue you were coming up against in the first place?

SteveMacenski replied:
Well, uncomment and see if it compiles (also need to uncomment the bit in the cmakelists), if it does try loading it through the component manager

SteveMacenski replied:
Also this ticket: https://github.com/SteveMacenski/ros2_ouster_drivers/issues/13

rotu reviewed (commented):
Can’t iterate on the code ATM since I’m on my phone - will get back to it soon

SteveMacenski commented:
No rush

rotu committed:
Code review comments

1. Fix order of fields in OS1 Sensor so default destructor destroys things in the right order, also make fields protected so subclasses can control order of destruction
2. Register OS1Driver as a component
3. Revert using ::make_unique with a node, which causes undefined behavior when shared (because it inherits from std::enable_shared_from_this)

rotu committed:
Revert unnecessary changes to OS1_sensor

These belong in another PR

SteveMacenski commented on ros2_ouster/src/ouster_driver.cpp:
```diff
@@ -11,15 +11,19 @@
 // See the License for the specific language governing permissions and
 // limitations under the License.
 
+#include "ros2_ouster/ouster_driver.hpp"
```
This needs to be below the STL headers with the other ouster headers

rotu replied:
Are you sure? That violates the Google coding standards, which ROS2 standards are based on.

> Include headers in the following order: Related header, C system headers, C++ standard library headers, other libraries' headers, your project's headers.

https://google.github.io/styleguide/cppguide.html#Names_and_Order_of_Includes

SteveMacenski replied:
Run the ROS linters, you'll find that any system includes `<>` before C++ includes `""` will trigger a linter issue.

rotu replied:
It passes the gauntlet in CI. Which linter is giving trouble? http://build.ros2.org/job/Epr__ros2_ouster_driver__ubuntu_bionic_amd64/27/consoleText

SteveMacenski replied:
Ah - that's strange. I don't know why I've had issues with this before. 

Either way, can you move it with the others please? If nothing else it is in line with the rest of the project's styling (and navigation, for that matter)

rotu replied:
I believe you and will run the linters locally when I bang on this next

SteveMacenski reviewed (commented):
Does the component register and such? (e.g. can you load it in to a component manager with other stuff?)

If so, I can live with this

rotu committed:
register nodes in cmakelists

rotu committed:
Restore strange include order

SteveMacenski reviewed (approved):
This looks good - did you test this on a sensor to make sure nothing breaks unexpectedly?

Also on the component side - does this get loaded in a manger OK?

I can test this too when I have to test #31 but that may have to wait until the weekend.

rotu commented:
No, I haven't tested this on a sensor. And I haven't loaded it in a manager. I don't see the code to do that in this repo anywhere.

SteveMacenski commented:
@rotu that's a ROS2 thing, not a specific program thing https://index.ros.org/doc/ros2/Tutorials/Composition/

rotu commented:
AFAICT it works:
```
$rros2 run rclcpp_components component_container
[INFO] [1586380760.635050661] [ComponentManager]: Load Library: /home/dan/Documents/ros_ouster_ws/install/lib/libouster_driver_core.so
[INFO] [1586380760.664994054] [ComponentManager]: Found class: rclcpp_components::NodeFactoryTemplate<ros2_ouster::OS1Driver>
[INFO] [1586380760.665054526] [ComponentManager]: Instantiate class: rclcpp_components::NodeFactoryTemplate<ros2_ouster::OS1Driver>
```

```
$ros2 component load /ComponentManager ros2_ouster ros2_ouster::OS1Driver
Loaded component 1 into '/ComponentManager' container node as '/OusterDriver'
```

SteveMacenski commented:
I tested this, works all fine, merging. Thanks for the work!
